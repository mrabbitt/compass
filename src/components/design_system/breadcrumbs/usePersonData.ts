import { trpc } from "@/client/lib/trpc";
import { SelectableForTable } from "zapatos/schema";

export type Student = SelectableForTable<"student">;
export type Para = SelectableForTable<"user">;
export type PersonData = Student | Para | undefined;

export const usePersonData = (query: Record<string, string>): PersonData => {
  const { data: student } = trpc.student.getStudentById.useQuery(
    { student_id: query.student_id },
    { enabled: query.student_id !== undefined }
  );
  const { data: para } = trpc.para.getParaById.useQuery(
    { user_id: query.user_id },
    { enabled: query.user_id !== undefined }
  );

  return student || para;
};
