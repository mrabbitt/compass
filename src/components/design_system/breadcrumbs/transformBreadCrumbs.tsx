import Link from "next/link";
import $breadcrumbs from "./Breadcrumbs.module.css";
import { PersonData } from "./usePersonData";

export const transformPaths = ({
  fullPath,
  personData,
}: {
  fullPath: string;
  personData: PersonData;
}) => {
  return fullPath.split("/").map((path, index) => {
    // 0th index seems to only be empty string
    if (index === 0) return "";
    // 1st index currently is either students or staff
    if (index % 2 === 1) {
      return (
        <Link key={index} href={`/${path}`} className={$breadcrumbs.link}>
          {
            path.toUpperCase() /* TODO ALL CAPS is generally bad for a11y; revisit this. */
          }
        </Link>
      );
    }
    // 2nd index is the ID referencing 1st index
    if (index === 2) {
      return (
        <div key={index} className={$breadcrumbs["non-link-crumb"]}>
          {personData?.first_name} {personData?.last_name}
        </div>
      );
    }
    return <div key={index}>{path}</div>;
  });
};
