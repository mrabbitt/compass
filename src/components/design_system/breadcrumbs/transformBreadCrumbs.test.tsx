/* eslint-disable react/jsx-key */
import "@testing-library/jest-dom";
import { describe, expect, jest, test } from "@jest/globals";
import { transformPaths } from "./transformBreadCrumbs";
import { Student } from "./usePersonData";

// TODO Consider having transformPaths return a list of objects that can be tested without snapshots, and handle rendering in BreadcrumbNav or elsewhere.
describe("transformPaths", () => {
  test("empty path", () => {
    const actual = transformPaths({ fullPath: "/", personData: undefined });

    expect(actual).toMatchSnapshot();
  });

  test("/students", () => {
    const actual = transformPaths({
      fullPath: "/students",
      personData: undefined,
    });

    expect(actual).toMatchSnapshot();
  });

  test("/students/student-id", () => {
    const actual = transformPaths({
      fullPath: "/students/student-id",
      personData: { first_name: "Alia", last_name: "Atreides" } as Student,
    });

    expect(actual).toMatchSnapshot();
  });

  test("/students/student-id/goals/goal-id", () => {
    const actual = transformPaths({
      fullPath: "/students/student-id/goals/goal-id",
      personData: { first_name: "Alia", last_name: "Atreides" } as Student,
    });

    expect(actual).toMatchSnapshot();
  });
});
