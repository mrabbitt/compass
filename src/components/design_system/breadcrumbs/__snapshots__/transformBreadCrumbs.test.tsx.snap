// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformPaths /students 1`] = `
[
  "",
  <ForwardRef(LinkComponent)
    className="link"
    href="/students"
  >
    STUDENTS
  </ForwardRef(LinkComponent)>,
]
`;

exports[`transformPaths /students/student-id 1`] = `
[
  "",
  <ForwardRef(LinkComponent)
    className="link"
    href="/students"
  >
    STUDENTS
  </ForwardRef(LinkComponent)>,
  <div
    className="non-link-crumb"
  >
    Alia
     
    Atreides
  </div>,
]
`;

exports[`transformPaths /students/student-id/goals/goal-id 1`] = `
[
  "",
  <ForwardRef(LinkComponent)
    className="link"
    href="/students"
  >
    STUDENTS
  </ForwardRef(LinkComponent)>,
  <div
    className="non-link-crumb"
  >
    Alia
     
    Atreides
  </div>,
  <ForwardRef(LinkComponent)
    className="link"
    href="/goals"
  >
    GOALS
  </ForwardRef(LinkComponent)>,
  <div>
    goal-id
  </div>,
]
`;

exports[`transformPaths empty path 1`] = `
[
  "",
  <ForwardRef(LinkComponent)
    className="link"
    href="/"
  >
    
  </ForwardRef(LinkComponent)>,
]
`;
